pkg/controller/component/storaged_updater.go

```diff
@@ -0,0 +1,320 @@
+	"github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1"
+	"github.com/vesoft-inc/nebula-operator/pkg/kube"
+	"github.com/vesoft-inc/nebula-operator/pkg/nebula"
+	utilerrors "github.com/vesoft-inc/nebula-operator/pkg/util/errors"
+	extenderutil "github.com/vesoft-inc/nebula-operator/pkg/util/extender"
+	"github.com/vesoft-inc/nebula-operator/pkg/util/resource"
+)
+
+const (
+	// TransLeaderBeginTime is the key of trans Leader begin time
+	TransLeaderBeginTime = "transLeaderBeginTime"
+	// TransLeaderTimeout is the timeout limit of trans leader
+	TransLeaderTimeout = 3 * time.Minute
+)
+
+type storagedUpdate struct {
+	client.Client
+	clientSet kube.ClientSet
+	extender  extenderutil.UnstructuredExtender
+}
+
+func NewStoragedUpdater(cli client.Client, clientSet kube.ClientSet) UpdateManager {
+	return &storagedUpdate{
+		Client:    cli,
+		clientSet: clientSet,
+		extender:  extenderutil.New(),
+	}
+}
+
+func (s *storagedUpdate) Update(
+	nc *v1alpha1.NebulaCluster,
+	oldUnstruct, newUnstruct *unstructured.Unstructured,
+	gvk schema.GroupVersionKind,
+) error {
+	log := getLog().WithValues("namespace", nc.GetNamespace(), "name", nc.GetName())
+	if *nc.Spec.Storaged.Replicas == int32(0) {
+		return nil
+	}
+
+	if nc.Status.Storaged.Phase == v1alpha1.ScaleInPhase || nc.Status.Storaged.Phase == v1alpha1.ScaleOutPhase {
+		log.Info("storaged is scaling, can not update")
+		return getLastConfig(s.extender, oldUnstruct, newUnstruct)
+	}
+
+	nc.Status.Storaged.Phase = v1alpha1.UpdatePhase
+	if !extenderutil.PodTemplateEqual(s.extender, newUnstruct, oldUnstruct) {
+		return nil
+	}
+
+	if nc.Status.Storaged.Workload.UpdateRevision == nc.Status.Storaged.Workload.CurrentRevision {
+		return nil
+	}
+
+	spec := s.extender.GetSpec(oldUnstruct)
+	actualStrategy := spec["updateStrategy"].(map[string]interface{})
+	advanced := gvk.Kind == resource.AdvancedStatefulSetKind.Kind
+	partition := actualStrategy["rollingUpdate"].(map[string]interface{})
+	if err := setPartition(s.extender, newUnstruct, partition["partition"].(int64), advanced); err != nil {
+		return err
+	}
+
+	endpoints := []string{nc.GetMetadThriftConnAddress()}
+	mc, err := nebula.NewMetaClient(endpoints)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := mc.Disconnect(); err != nil {
+			log.Error(err, "meta client disconnect failed")
+		}
+	}()
+
+	replicas := s.extender.GetReplicas(oldUnstruct)
+	index, err := getNextUpdatePod(nc.StoragedComponent(), *replicas, s.clientSet.Pod())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return utilerrors.ReconcileErrorf("%v", err)
+		}
+		return err
+	}
+	if index >= 0 {
+		return s.updateStoragedPod(mc, nc, index, newUnstruct, advanced)
+	}
+
+	if err := mc.BalanceLeader(); err != nil {
+		log.Error(err, "balance leader failed")
+		return err
+	}
+
+	hostItem, err := mc.ListHosts()
+	if err != nil {
+		return err
+	}
+	if !mc.IsBalanced(hostItem) {
+		if err := mc.BalanceLeader(); err != nil {
```

Comment:
It has been done before, and why `BalanceLeader` again?

Reply:
Here is rechecking